infra/Pos/Slotting/Impl/Util.hs

```diff
@@ -17,55 +17,79 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
                                               mkLocalSlotIndex)
 import           Pos.Util.Util               (leftToPanic)
 
+import           Pos.Slotting.Util           (getEpochSlottingDataEmpatically)
 import           Pos.Slotting.MemState.Class (MonadSlotsData (..))
-import           Pos.Slotting.Types          (EpochSlottingData (..), SlottingData (..))
+import           Pos.Slotting.Types          (EpochSlottingData (..))
 
 -- | Approximate current slot using outdated slotting data.
-approxSlotUsingOutdated :: MonadSlotsData m => EpochIndex -> Timestamp -> m SlotId
+approxSlotUsingOutdated
+    :: (MonadSlotsData m)
+    => EpochIndex
+    -> Timestamp
+    -> m SlotId
 approxSlotUsingOutdated penult t = do
-    SlottingData {..} <- getSlottingData
-    systemStart <- getSystemStart
+    systemStart <- getSystemStartM
+    sdLast      <- getNextEpochSlottingDataM
     let epochStart = esdStartDiff sdLast `addTimeDiffToTimestamp` systemStart
     pure $
         if | t < epochStart -> SlotId (penult + 1) minBound
            | otherwise      -> outdatedEpoch systemStart t (penult + 1) sdLast
   where
     outdatedEpoch systemStart (Timestamp curTime) epoch EpochSlottingData {..} =
         let duration = convertUnit esdSlotDuration
-            start = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart) in
+            start = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart)
+        in
         unflattenSlotId $
         flattenEpochIndex epoch + fromIntegral ((curTime - start) `div` duration)
 
+-- TODO(ks): Move to MonadSlotsData?
 -- | Compute current slot from current timestamp based on data
 -- provided by 'MonadSlotsData'.
 slotFromTimestamp
-    :: MonadSlotsData m
-    => Timestamp -> m (Maybe SlotId)
+    :: (MonadSlotsData m, MonadThrow m)
+    => Timestamp
+    -> m (Maybe SlotId)
 slotFromTimestamp approxCurTime = do
-    SlottingData {..} <- getSlottingData
-    systemStart <- getSystemStart
-    let tryEpoch = computeSlotUsingEpoch systemStart approxCurTime
-    let penultRes = tryEpoch sdPenultEpoch sdPenult
-    let lastRes = tryEpoch (succ sdPenultEpoch) sdLast
-    return $ penultRes <|> lastRes
 
-computeSlotUsingEpoch
-    :: Timestamp
-    -> Timestamp
-    -> EpochIndex
-    -> EpochSlottingData
-    -> Maybe SlotId
-computeSlotUsingEpoch systemStart (Timestamp curTime) epoch EpochSlottingData {..}
-    | curTime < epochStart = Nothing
-    | curTime < epochStart + epochDuration = Just $ SlotId epoch localSlot
-    | otherwise = Nothing
+    systemStart     <- getSystemStartM
+    allEpochIndex   <- getAllEpochIndexM
+
+    -- We partially apply the function and then iterate over all epoch indexes.
+    mSlotId         <- forM allEpochIndex (findSlot systemStart approxCurTime)
+
+    pure $ head $ catMaybes mSlotId
```

Comment:
Well, to be honest, I am afraid that it can be quite inefficient. Especially because you start from the oldest epochs. This function is called quite often and the size of `SlottingData` can will become bigger and bigger.
And this code will traverse all epochs from the 0-th to the _next_ one and do some computations. I propose to do two modifications:
1. Start from the newest epoch.
2. Stop if the result is computed successfully (so it should be `foldM` most likely).